File: backend/KtorApp/src/main/kotlin/com/example/utils/BookProcessor.kt
package com.example.utils

import com.example.models.Book
import com.example.models.BookText
import com.example.models.TextBlock
import org.apache.pdfbox.pdmodel.PDDocument
import org.apache.pdfbox.text.PDFTextStripper
import org.bson.types.ObjectId
import java.io.File

class BookProcessor(private val filePath: String, private val nameFile: String) {

    /**
     * Чтение текстового файла.
     */
    fun readTextFile(): String {
        return File(filePath).readText()
    }

    /**
     * Чтение PDF-файла и извлечение текста.
     */
    fun readPdfFile(): String {
        PDDocument.load(File(filePath)).use { document ->
            val pdfStripper = PDFTextStripper()
            return pdfStripper.getText(document) // Используем метод getText(document)
        }
    }

    /**
     * Чтение FB2-файла и извлечение структурированных данных.
     */
    fun readFb2File(): BookText {
        val fb2Processor = Fb2Processor(filePath)
        return fb2Processor.extractBookText()
    }

    /**
     * Получение списка глав из содержимого книги.
     */
    fun getChapters(fileType: String): List<String> {
        val content = when (fileType) {
            "text/plain" -> readTextFile()
            "application/pdf" -> readPdfFile()
            "application/fb2+xml" -> readFb2File().content
            else -> throw IllegalArgumentException("Unsupported file type")
        }
        return content.split(Regex("(Глава|Часть)\\s+\\d+", RegexOption.IGNORE_CASE))
            .filter { it.isNotBlank() }
    }

    /**
     * Создание объекта Book на основе типа файла.
     */
    fun getBook(fileType: String): Book {
        val bookText = when (fileType) {
            "text/plain" -> {
                val content = readTextFile()
                BookText(
                    title = extractLineContent(content, "Title:") ?: "Название не указано",
                    authors = extractLineContent(content, "Author(s):") ?: "Автор книги не указан",
                    content = extractLineContent(content, "Content:") ?: "Описание отсутствует",
                    annotation = null // Текстовые файлы, возможно, не содержат аннотацию
                )
            }
            "application/pdf" -> {
                val text = readPdfFile()
                // Предполагается, что PDF обрабатывается аналогично FB2
                BookText(
                    title = "Неизвестно", // PDF-парсер не извлекает метаданные
                    authors = "Неизвестно",
                    content = text,
                    annotation = null // PDF-файлы, возможно, не содержат аннотацию
                )
            }
            "application/fb2+xml" -> readFb2File()
            else -> throw IllegalArgumentException("Unsupported file type")
        }

        val idBook = ObjectId().toString()

        return Book(
            idBook = idBook, // Здесь idBook создается
            title = bookText.title,
            author = bookText.authors,
            description = bookText.content,
            annotation = bookText.annotation,
            status = "reading",
            mode = "default",
            nameFile = nameFile,
            filePath = filePath
        )
    }

    /**
     * Вспомогательный метод для извлечения содержимого после заданного префикса.
     */
    private fun extractLineContent(content: String, prefix: String): String? {
        val regex = Regex("(?i)$prefix\\s*(.*)")
        val matchResult = regex.find(content)
        return matchResult?.groups?.get(1)?.value?.trim()
    }

    /**
     * Разделение главы на блоки по 1000 символов.
     */
    fun divideChapterIntoBlocks(chapter: String): List<String> {
        val blockSize = 1000
        val blocks = mutableListOf<String>()
        var startIndex = 0

        while (startIndex < chapter.length) {
            val endIndex = (startIndex + blockSize).coerceAtMost(chapter.length)
            blocks.add(chapter.substring(startIndex, endIndex))
            startIndex = endIndex
        }

        return blocks
    }

    /**
     * Обработка глав и разделение их на текстовые блоки.
     */
    fun processChaptersAndBlocks(chapters: List<String>): List<TextBlock> {
        val textBlocks = mutableListOf<TextBlock>()
        chapters.forEachIndexed { chapterIndex, chapterContent ->
            val blocks = divideChapterIntoBlocks(chapterContent)
            blocks.forEach { blockContent ->
                val textBlock = TextBlock(
                    _id = ObjectId().toString(), // Используем _id вместо idBlock
                    original = blockContent,
                    numberChapter = chapterIndex + 1
                )
                textBlocks.add(textBlock)
            }
        }
        return textBlocks
    }
}

File: backend/Utils/BookProcessor.py
import re
from typing import List, Optional
from bson import ObjectId
from PyPDF2 import PdfReader
from models.BookText import BookText
from models.Book import Book
from models.TextBlock import TextBlock
from Utils.Fb2Processor import Fb2Processor  # Предполагается, что вы уже адаптировали Fb2Processor

class BookProcessor:
    def __init__(self, file_path: str, name_file: str):
        self.file_path = file_path
        self.name_file = name_file

    def read_text_file(self) -> str:
        with open(self.file_path, 'r', encoding='utf-8', errors='ignore') as f:
            return f.read()

    def read_pdf_file(self) -> str:
        text_content = []
        with open(self.file_path, 'rb') as f:
            reader = PdfReader(f)
            for page in reader.pages:
                page_text = page.extract_text()
                if page_text:
                    text_content.append(page_text)
        return "\n".join(text_content)

    def read_fb2_file(self) -> BookText:
        fb2_processor = Fb2Processor(self.file_path)
        return fb2_processor.extract_book_text()

    def get_chapters(self, file_type: str) -> List[str]:
        if file_type == "text/plain":
            content = self.read_text_file()
        elif file_type == "application/pdf":
            content = self.read_pdf_file()
        elif file_type == "application/fb2+xml":
            content = self.read_fb2_file().content
        else:
            raise ValueError("Unsupported file type")

        # Разделяем на главы по шаблону "(Глава|Часть)\s+\d+"
        chapters = re.split(r"(Глава|Часть)\s+\d+", content, flags=re.IGNORECASE)
        # Фильтруем пустые строки и пробелы
        chapters = [ch.strip() for ch in chapters if ch.strip()]
        return chapters

    def get_book(self, file_type: str) -> Book:
        if file_type == "text/plain":
            content = self.read_text_file()
            book_text = BookText(
                title=self.extract_line_content(content, "Title:") or "Название не указано",
                authors=self.extract_line_content(content, "Author(s):") or "Автор книги не указан",
                content=self.extract_line_content(content, "Content:") or "Описание отсутствует",
                annotation=None
            )
        elif file_type == "application/pdf":
            text = self.read_pdf_file()
            book_text = BookText(
                title="Неизвестно",
                authors="Неизвестно",
                content=text,
                annotation=None
            )
        elif file_type == "application/fb2+xml":
            book_text = self.read_fb2_file()
        else:
            raise ValueError("Unsupported file type")

        id_book = str(ObjectId())
        return Book(
        idBook=id_book,
        title=book_text.title,
        author=book_text.authors,
        description=book_text.content,
        annotation=book_text.annotation,
        status="reading",
        mode="default",
        nameFile=self.name_file,
        filePath=self.file_path,
        compressionLevel=0
    )

    def extract_line_content(self, content: str, prefix: str) -> Optional[str]:
        pattern = re.compile(prefix + r"\s*(.*)", re.IGNORECASE)
        match = pattern.search(content)
        if match:
            return match.group(1).strip()
        return None

    def divide_chapter_into_blocks(self, chapter: str, block_size: int = 1000) -> List[str]:
        blocks = []
        start_index = 0
        while start_index < len(chapter):
            end_index = min(start_index + block_size, len(chapter))
            blocks.append(chapter[start_index:end_index])
            start_index = end_index
        return blocks

    def process_chapters_and_blocks(self, chapters: List[str]) -> List[TextBlock]:
        text_blocks = []
        for chapter_index, chapter_content in enumerate(chapters):
            blocks = self.divide_chapter_into_blocks(chapter_content)
            for block_content in blocks:
                text_block = TextBlock(
                    _id=str(ObjectId()),
                    original=block_content,
                    numberChapter=chapter_index + 1
                )
                text_blocks.append(text_block)
        return text_blocks
